### PR TITLE
Movie cleanup 2021

### PR DESCRIPTION
**Pull request by TheCycoONE**
Drops support for FFMPEG 3.0 and lower.
Drops support for LibAV (any version)
Less raw pointer use, and in particular raw owned pointers
No more implicit casts
Reworked audio conversion (no more audio buffer)
Replaced linked list implementation in packet queue with std::queue
Fix for possible undefined behavior around aborting

### CODE REVIEW

File: CorsixTH/Src/th_movie.cpp [hunk 1]
```diff
@@ -296,59 +247,28 @@ int movie_picture_buffer::write(AVFrame* pFrame, double dPts) {
...
-    pPacket = (AVPacket*)av_malloc(sizeof(AVPacket));
-    *pPacket = pNode->pkt;
-    av_free(pNode);
+  av_packet_unique_ptr pPacket(nullptr);
+  if (!data.empty()) {
+    pPacket.swap(data.front());
+    data.pop();
   }
 
   return pPacket;
```
Comment (Alberth289346) [inline CorsixTH/Src/th_movie.cpp:274]: Old problem, but this looks weird. It locks the mutex at the start of the function but doesn't release it here.

Comment (TheCycoONE) [reply CorsixTH/Src/th_movie.cpp:274]: Yes it does. C++ mutexes release at end of scope.

**Comment by Alberth289346** [comment]
LGTM

**Comment by TheCycoONE** [comment]
Still in progress, hence draft, but good so far 👍

File: CorsixTH/Src/th_movie.h [hunk 2]
```diff
@@ -411,22 +423,22 @@ class movie_player {
...
   int audio_buffer_size;      ///< The current size of audio data in
-                              ///< #m_pbAudioBuffer
+                              ///< #audio_buffer
   int audio_buffer_index;     ///< The current position for writing in
-                              ///< #m_pbAudioBuffer
-  int audio_buffer_max_size;  ///< The capacity of #m_pbAudioBuffer (allocated
+                              ///< #audio_buffer
+  int audio_buffer_max_size;  ///< The capacity of #audio_buffer (allocated
                               ///< size)
   uint8_t* audio_buffer;      ///< An audio buffer for playback
```
Comment (Alberth289346) [inline CorsixTH/Src/th_movie.h:431]: an class with data storage and these administrative numbers could be useful?

Comment (TheCycoONE) [reply CorsixTH/Src/th_movie.h:449]: std::array/std::vector can be used for that; will get around to replacing these in this patch set.

**Comment by TheCycoONE** [comment]
I think that's just about enough things for one patch set.
Best to review each commit independently rather than the whole set - I mostly tried to keep each step compartmentalized.

File: CorsixTH/Src/th_movie.h [hunk 3]
```diff
@@ -169,28 +183,40 @@ class movie_picture_buffer {
...
+  //! should only be accessed on the main thread
+  SDL_Texture* texture;
+
+  //! A mutex for restricting access to the picture buffer to a single thread
+  std::mutex mutex;
+
+  //! A condition for indicating access to the picture buffer
+  std::condition_variable cond;
+
+  //! The picture queue, aloo ping FIFO queue of movie_pictures
```
Comment (TheCycoONE) [inline CorsixTH/Src/th_movie.h:218]: aloo ping!

File: CorsixTH/Src/th_movie.cpp [hunk 4]
```diff
@@ -729,17 +693,11 @@ int movie_player::decode_audio_frame(bool fFirst) {
...
-#elif defined(CORSIX_TH_USE_LIBAV)
-  avresample_convert(audio_resample_context, &audio_buffer, 0, iOutSamples,
-                     (uint8_t**)&audio_frame->data[0], 0,
-                     audio_frame->nb_samples);
-#endif
   return iSampleSize;
 }
-#else   // CORSIX_TH_USE_FFMPEG || CORSIX_TH_USE_LIBAV
+#else   // CORSIX_TH_USE_FFMPEG
 movie_player::movie_player() {}
```
Comment (Alberth289346) [inline CorsixTH/Src/th_movie.cpp:681]: We seem to indent code for `#` lines too, but not here apprarently?

Comment (Alberth289346) [reply CorsixTH/Src/th_movie.cpp:681]: `th_movie.h` also doesn't always, maybe we have another rule?

**Comment by Alberth289346** [comment]
lgtm

**Pull request merged**